lib/Frontend/ParseableInterfaceModuleLoader.cpp

```diff
@@ -0,0 +1,1015 @@
+    if (!fwdBuf) return false;
+    llvm::yaml::Input yamlIn(fwdBuf.get()->getBuffer());
+    yamlIn >> forwardingModule;
+    hasLoadedForwardingModule = true;
+    return true;
+  }
+
+  const ForwardingModule &getForwardingModule() const {
+    assert(kind == Kind::Forwarding &&
+           "getForwardingModule() called on non-forwarding module");
+    assert(hasLoadedForwardingModule &&
+           "must loading forwarding module contents before accessing");
+    return forwardingModule;
+  }
+
+  // Returns the path to the .swiftmodule that will eventually be loaded by
+  // this
+  StringRef getModulePath() const {
+    switch (kind) {
+      case Kind::Normal:
+      case Kind::Prebuilt:
+        return path;
+      case Kind::Forwarding:
+        return getForwardingModule().underlyingModulePath;
+    }
+  }
+};
+
+} // end anonymous namespace
+
+#pragma mark - Common utilities
+
+namespace path = llvm::sys::path;
+
+static std::unique_ptr<llvm::MemoryBuffer>
+getASTBuffer(llvm::vfs::FileSystem &fs, StringRef path) {
+  auto ModBuf = fs.getBufferForFile(path, /*FileSize=*/-1,
+                                    /*RequiresNullTerminator=*/false);
+  return ModBuf ? std::move(*ModBuf) : nullptr;
+}
+
+static bool serializedASTLooksValid(const llvm::MemoryBuffer &buf) {
+  auto VI = serialization::validateSerializedAST(buf.getBuffer());
+  return VI.status == serialization::Status::Valid;
+}
+
+static std::unique_ptr<llvm::MemoryBuffer> getBufferOfDependency(
+  llvm::vfs::FileSystem &fs, StringRef depPath, StringRef interfacePath,
+  DiagnosticEngine &diags, SourceLoc diagnosticLoc) {
+  auto depBuf = fs.getBufferForFile(depPath, /*FileSize=*/-1,
+                                    /*RequiresNullTerminator=*/false);
+  if (!depBuf) {
+    diags.diagnose(diagnosticLoc,
+                   diag::missing_dependency_of_parseable_module_interface,
+                   depPath, interfacePath, depBuf.getError().message());
+    return nullptr;
+  }
+  return std::move(depBuf.get());
+}
+
+static Optional<llvm::vfs::Status> getStatusOfDependency(
+  llvm::vfs::FileSystem &fs, StringRef depPath, StringRef interfacePath,
+  DiagnosticEngine &diags, SourceLoc diagnosticLoc) {
+  auto status = fs.status(depPath);
+  if (!status) {
+    diags.diagnose(diagnosticLoc,
+                   diag::missing_dependency_of_parseable_module_interface,
+                   depPath, interfacePath, status.getError().message());
+    return None;
+  }
+  return status.get();
+}
+
+#pragma mark - Module Building
+
+// needs to be in the swift namespace so CompilerInvocation can see it
+namespace swift {
+
+class ParseableInterfaceBuilder {
+  ASTContext &ctx;
+  llvm::vfs::FileSystem &fs;
+  DiagnosticEngine &diags;
+  StringRef interfacePath;
+  StringRef moduleName;
+  StringRef moduleCachePath;
+  StringRef prebuiltCachePath;
+  bool serializeDependencyHashes;
+  SourceLoc diagnosticLoc;
+  DependencyTracker *const dependencyTracker;
```

Comment:
Why is just this member `const` and not any others?

Reply:
You're right, actually most of these should be const.